privval/signer_listener_endpoint.go

```diff
@@ -0,0 +1,294 @@
+	if err != nil {
+		sl.Logger.Debug("Read Error", "err", err)
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (sl *SignerListenerEndpoint) isConnected() bool {
+	return sl.IsRunning() && sl.conn != nil
+}
+
+func (sl *SignerListenerEndpoint) readMessage() (msg RemoteSignerMsg, err error) {
+	if !sl.isConnected() {
+		return nil, cmn.ErrorWrap(ErrListenerNoConnection, "endpoint is not connected")
+	}
+
+	// Reset read deadline
+	deadline := time.Now().Add(sl.timeoutReadWrite)
+	sl.Logger.Debug(
+		"SignerListenerEndpoint: readMessage",
+		"timeout", sl.timeoutReadWrite,
+		"deadline", deadline)
+
+	err = sl.conn.SetReadDeadline(deadline)
+	if err != nil {
+		return
+	}
+
+	const maxRemoteSignerMsgSize = 1024 * 10
+	_, err = cdc.UnmarshalBinaryLengthPrefixedReader(sl.conn, &msg, maxRemoteSignerMsgSize)
+	if _, ok := err.(timeoutError); ok {
+		err = cmn.ErrorWrap(ErrListenerTimeout, err.Error())
+		sl.dropConnection()
+	}
+
+	return
+}
+
+func (sl *SignerListenerEndpoint) writeMessage(msg RemoteSignerMsg) (err error) {
+	if !sl.isConnected() {
+		return cmn.ErrorWrap(ErrListenerNoConnection, "endpoint is not connected")
+	}
+
+	// Reset read deadline
+	deadline := time.Now().Add(sl.timeoutReadWrite)
+	sl.Logger.Debug(
+		"SignerListenerEndpoint: writeMessage",
+		"timeout", sl.timeoutReadWrite,
+		"deadline", deadline)
+
+	err = sl.conn.SetWriteDeadline(deadline)
+	if err != nil {
+		return
+	}
+
+	_, err = cdc.MarshalBinaryLengthPrefixedWriter(sl.conn, msg)
+	if _, ok := err.(timeoutError); ok {
+		err = cmn.ErrorWrap(ErrListenerTimeout, err.Error())
+	}
+
+	return
+}
+
+func (sl *SignerListenerEndpoint) ensureConnection(maxWait time.Duration) error {
+	if !sl.isConnected() {
+		// Is there a connection ready?
+		select {
+		case sl.conn = <-sl.connectedCh:
+			{
+				sl.Logger.Debug("SignerListenerEndpoint: received connection")
+				return nil
+			}
+		default:
+			{
+				sl.Logger.Debug("SignerListenerEndpoint: no connection is ready")
+			}
+		}
+
+		// should we trigger a reconnect?
+		select {
+		case sl.connectCh <- struct{}{}:
+			{
+				sl.Logger.Debug("SignerListenerEndpoint: triggered a reconnect")
+			}
+		default:
+			{
+				sl.Logger.Debug("SignerListenerEndpoint: reconnect in progress")
+			}
+		}
+
+		// block until connected or timeout
+		select {
+		case sl.conn = <-sl.connectedCh:
+			{
+				sl.Logger.Debug("SignerListenerEndpoint: connected")
+			}
+		case <-time.After(maxWait):
+			{
+				sl.Logger.Debug("SignerListenerEndpoint: timeout")
+				return ErrListenerTimeout
+			}
+		}
+	}
+	return nil
+}
+
+func (sl *SignerListenerEndpoint) dropConnection() {
+	if sl.conn != nil {
+		if err := sl.conn.Close(); err != nil {
+			sl.Logger.Error("SignerListenerEndpoint::dropConnection", "err", err)
+		}
+		sl.conn = nil
+	}
+}
+
+func (sl *SignerListenerEndpoint) serviceLoop() {
+	defer close(sl.stoppedCh)
+	sl.Logger.Debug("SignerListenerEndpoint::serviceLoop")
+
+	for {
+		select {
+		case <-sl.connectCh:
+			{
+				for {
```

Comment:
this infinite loop within another infinite loop looks weird (can be problematic in the future)

Reply:
agree, we can probably refactor this a bit more